### PR TITLE
All the template parameters!!

### PR DESCRIPTION
**Pull request by fitzgen**
The major changes in this PR are:

* De-duplication of named template type parameters (this is probably the biggest part of the PR, and nastiest because it is the part that deals with libclang)

* Removing the `signature_contains_named_type` stuff, and enabling the more sound template type parameter usage analysis that has been landing in bits here and there.

* **LOTS** of new tests for template type parameter usage

@emilio: can you also test this on the stylo bindings? I tested on the SpiderMonkey bindings and found a bug and fixed a bug related to instantiations of partially specialized templates. I'd like to make sure that there aren't any latent bugs uncovered in Stylo.

This is NOT ready to merge quite yet, but is ready for some more eyeballs that are not mine.

Still TODO:

* [x] Rebase so that these changes will merge cleanly -- I'll get on this ASAP

* [x] Time SpiderMonkey bindings generation with and without these changes to see what the overhead of the new analysis is (if any!) on Large and Real World bindings

* [x] Test theses changes on Stylo (thanks @emilio!)

    * [ ] (optional) and time Stylo bindings generation with and without these changes as well (only if you want to, @emilio)

Thanks!

### CODE REVIEW

File: src/ir/context.rs [hunk 1]
```diff
@@ -569,6 +569,24 @@ impl<'ctx> BindgenContext<'ctx> {
...
     }
 
+    /// Return `true` if `item` uses the given `template_param`, `false`
+    /// otherwise.
+    ///
+    /// This method may only be called during the codegen phase, because the
+    /// template usage information is only computed as we enter the codegen
+    /// phase.
+    pub fn uses_template_parameter(&self, item: ItemId, template_param: ItemId) -> bool {
+        assert!(self.in_codegen_phase(),
```
Comment (emilio) [inline src/ir/context.rs]: Perhaps we should move this to something like a `Phase` enum?

```rust
enum Phase {
    Parsing,
    ReplacementProcessing,
    Codegen,
}
```

With something like:

```rust
impl BindgenContext {
    pub fn advance_phase(&mut self, to: Phase) {
        assert!(to > self.current_phase);
        self.current_phase = to;
    }
}
```

And make more fine-grained assertions?

It's probably fine as a followup, or even an easy bug. Do you think it's worth filing?

File: src/ir/ty.rs [hunk 2]
```diff
@@ -101,6 +129,25 @@ pub trait TemplateDeclaration {
                  .collect())
         }
     }
+
+    /// Get only the set of template parameters that this item uses. This is a
+    /// subset of `all_template_params` and does not necessarily contain any of
+    /// `self_template_params`.
+    fn used_template_params(&self, ctx: &BindgenContext) -> Option<Vec<ItemId>>
```
Comment (emilio) [inline src/ir/ty.rs]: Is the option needed? I believe in most cases it'd be more convenient to just return a `Vec<ItemId>`, and make an empty vector represent the lack of them?

**Comment by fitzgen** [comment]
Rebased!

File: src/ir/item.rs [hunk 3]
```diff
@@ -346,6 +366,10 @@ pub struct Item {
     parent_id: ItemId,
     /// The item kind.
     kind: ItemKind,
+    /// Detect cycles when determining if we can derive debug/copy or not, and
+    /// avoid infinite recursion.
+    detect_derive_debug_cycle: Cell<bool>,
```
Comment (emilio) [inline src/ir/item.rs]: Can we remove the relevant fields from `CompInfo`? If not, why?

Is it worth to add an `struct AutoRestore<T: Copy>(&'a Cell<T>)` to simplify the code flow here? If so, mind filing an issue for it?

File: src/ir/ty.rs [hunk 4]
```diff
@@ -901,6 +901,13 @@ impl Type {
             _ => {}
         }
 
+        if location.kind() == CXCursor_ClassTemplatePartialSpecialization {
+            // Sorry! (Not sorry)
+            warn!("Found a partial template specialization; bindgen does not \
```
Comment (emilio) [inline src/ir/ty.rs]: Perhaps printing the location in the message is useful?

File: src/ir/context.rs (hunk 1 above)
Comment (fitzgen) [inline src/ir/context.rs]: I think even better would be to have different `Context` types and get rid of that nasty transmute. Then we could simply have methods that require more info on the context that has that info by construction, and push all these asserts into the type system at compile time.

File: src/ir/comp.rs [hunk 5]
```diff
@@ -35,7 +35,7 @@ pub enum MethodKind {
 }
 
 /// A struct representing a C++ method, either static, normal, or virtual.
-#[derive(Debug)]
+#[derive(Debug, PartialEq)]
```
Comment (emilio) [inline src/ir/comp.rs]: I guess I'll figure out why later, but I'm curious. The point of not doing this is that it's both expensive and not quite accurate, given pointing to different `ItemId` doesn't necessarily mean the two `ItemId`s are unequal.

File: src/ir/context.rs [hunk 6]
```diff
@@ -319,7 +319,8 @@ impl<'ctx> BindgenContext<'ctx> {
                item,
                definition);
 
-        assert!(item.is_named());
+        assert!(*item.as_type().unwrap().kind() == TypeKind::Named,
```
Comment (emilio) [inline src/ir/context.rs]: nit: `item.expect_type().kind().is_named()`?

If the `PartialEq` commit is due to this assert, I'd rather remove it and add a convenience method to `TypeKind`, or use the `matches` crate.

File: src/ir/template.rs [hunk 7]
```diff
@@ -34,6 +34,22 @@ use super::derive::CanDeriveCopy;
 use super::item::Item;
 use super::traversal::{EdgeKind, Trace, Tracer};
 
+/// A trait for things which may or may not be a named template type parameter.
+pub trait AsNamed {
```
Comment (emilio) [inline src/ir/template.rs]: I'd probably rename this to `AsNamedType`, what do you think?

**Comment by fitzgen** [comment]
Looks like I need to disable the last commit's new test on libclang <= 3.8:

```diff
 #[repr(C)]
 #[derive(Debug, Default, Copy)]
 pub struct Usage {
     pub _address: u8,
 }
 extern "C" {
     #[link_name = "_ZN5Usage13static_memberE"]
     pub static mut Usage_static_member: [u32; 2usize];
 }
 #[test]
 fn bindgen_test_layout_Usage() {
     assert_eq!(::std::mem::size_of::<Usage>() , 1usize , concat ! (
                "Size of: " , stringify ! ( Usage ) ));
     assert_eq! (::std::mem::align_of::<Usage>() , 1usize , concat ! (
                 "Alignment of " , stringify ! ( Usage ) ));
 }
+extern "C" {
+    #[link_name = "_ZN5UsageC1Ev"]
+    pub fn Usage_Usage(this: *mut Usage);
+}
 impl Clone for Usage {
     fn clone(&self) -> Self { *self }
+}
+impl Usage {
+    #[inline]
+    pub unsafe fn new() -> Self {
+        let mut __bindgen_tmp = ::std::mem::uninitialized();
+        Usage_Usage(&mut __bindgen_tmp);
+        __bindgen_tmp
+    }
 }
thread 'header_partial_specialization_and_inheritance_hpp' panicked at 'Header and binding differ!', /tmp/bindgen/debug/build/bindgen-e852d23f6698dc85/out/tests.rs:210
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

File: tests/expectations/lib.rs [hunk 8]
```diff
@@ -0,0 +1,3 @@
+#![allow(dead_code)]
+#![allow(non_camel_case_types)]
```
Comment (emilio) [inline tests/expectations/lib.rs]: This fixes an issue somewhere right?

**Comment by fitzgen** [comment]
(It the lack of seeing that the constructor is inline, but we need the constructor to trigger the panic that I was reducing when I generated this test case)

File: src/ir/ty.rs (hunk 2 above)
Comment (fitzgen) [inline src/ir/ty.rs]: The idea is to differentiate between these cases:

```
template <> class IsTemplateWithNoParams {};

class IsNotTemplate {};
```

I don't know if it is worth it however, as we eventually treat them identically.

File: src/ir/named.rs [hunk 9]
```diff
@@ -257,7 +257,7 @@ pub struct UsedTemplateParameters<'ctx, 'gen>
     where 'gen: 'ctx
 {
     ctx: &'ctx BindgenContext<'gen>,
-    used: HashMap<ItemId, ItemSet>,
+    used: HashMap<ItemId, Option<ItemSet>>,
```
Comment (emilio) [inline src/ir/named.rs]: Perhaps a comment here saying that the `Option` here is just for moving out of the hashmap and in again would be helpful.

File: src/ir/named.rs [hunk 10]
```diff
@@ -359,33 +359,47 @@ impl<'ctx, 'gen> MonotoneFramework for UsedTemplateParameters<'ctx, 'gen>  {
     }
 
     fn constrain(&mut self, id: ItemId) -> bool {
-        let original_len = self.used[&id].len();
+        // Invariant: all hash map entries' values are `Some` upon entering and
+        // exiting this method.
+        debug_assert!(self.used.values().all(|v| v.is_some()));
```
Comment (emilio) [inline src/ir/named.rs]: This can get quite hot I guess, could we maybe running it only on testing or something?

Comment (emilio) [inline src/ir/named.rs]: (Maybe it's not such an issue)

**Comment by fitzgen** [comment]
# Timing SpiderMonkey Bindings Generation

time | master | with these changes
---- | ------ | ------------------
real | 9.50   | 9.24
user | 9.35   | 9.10
sys  | 0.11   | 0.10

File: src/ir/ty.rs (hunk 2 above)
Comment (emilio) [inline src/ir/ty.rs]: I guess that's fair, maybe worth mentioning this bit in the comment?

**Comment by fitzgen** [comment]
So it looks like we are spending about the same amount of time on bindings generation. If I keep regenerating bindings, the numbers fluctuate a bit, but are about the same in either case. Sometimes master is faster than with this PR, sometimes the other way around. Doesn't look like this PR has much of an impact on performance either way.

File: src/ir/context.rs (hunk 1 above)
Comment (emilio) [inline src/ir/context.rs]: That sounds fine, though it's obviously a bigger change ;)

**Comment by emilio** [comment]
> So it looks like we are spending about the same amount of time on bindings generation. If I keep regenerating bindings, the numbers fluctuate a bit, but are about the same in either case. Sometimes master is faster than with this PR, sometimes the other way around. Doesn't look like this PR has much of an impact on performance either way.

That sounds fine! I'd run it on stylo when I have the chance :)

File: src/ir/ty.rs (hunk 2 above)
Comment (fitzgen) [inline src/ir/ty.rs]: Can do, will wait for your complete review before I start changing anything.

File: src/ir/context.rs (hunk 1 above)
Comment (fitzgen) [inline src/ir/context.rs]: Yeah, but also something we need pretty badly IMO. The transmuted lifetime can make it easy to accidentally have conflicting lifetimes for no good reason. I spent probably a half hour to an hour trying to get the named template parameter analysis working with the made up lifetime :-/

**Comment by fitzgen** [comment]
r? @emilio :)

This is ready for a real review now (assumign that you can use it to generate stylo bindings successfully)

**Comment by fitzgen** [comment]
> My main concern is that the actual unused template parameter can be pretty useful when writing abstractions on top of the bindgen-generated code.

This is a bit unfortunate. However, it is also the kind of thing where leveraging these templates on the Rust side already requires a bit of wrapping in a Rust-y API, and so a little bit more of that wrapping (adding a phantom type and whatever methods) doesn't seem like the end of the world. The trade off we're getting is more correctness, which almost always seems like a good trade off.

> It seems (from the reviewed code experience, which might not be so much) that switching from this code generation to the alternative (what we have now, template params + PhantomData) shouldn't be extremely hard (of course, it'd bring back the issues we have with template aliases). Is that right?

We could add an annotation that is like "please consider all my template parameters used" and it should be pretty easy. Unless it is crucial for Stylo, I'd like to do such things as a follow up.

File: src/ir/context.rs (hunk 6 above)
Comment (fitzgen) [inline src/ir/context.rs]: > If the PartialEq commit is due to this assert, I'd rather remove it and add a convenience method to TypeKind

SGTM

File: src/ir/context.rs [hunk 11]
```diff
@@ -103,6 +103,10 @@ pub struct BindgenContext<'ctx> {
     /// item ids during parsing.
     types: HashMap<TypeKey, ItemId>,
 
+    /// Maps from a cursor to the item id of the named template type parameter
+    /// for that cursor.
+    named_types: HashMap<clang::Cursor, ItemId>,
```
Comment (fitzgen) [inline src/ir/context.rs]: Seems like a good follow up. We also need to test whether it has any observable effect.

File: src/ir/named.rs (hunk 10 above)
Comment (fitzgen) [inline src/ir/named.rs]: It will only run in `debug` builds, which I don't think anyone is using outside hacking on bindgen itself, so I think we're fine.

Also, these sets are very very rarely even as big as 10 elements.

File: src/ir/item.rs (hunk 3 above)
Comment (fitzgen) [inline src/ir/item.rs]: I think the real fix is https://github.com/servo/rust-bindgen/issues/536. That will not only remove the need for these cells, but also make it so we can generate this stuff for everything all at once rather than re-walking the same sub graphs repeatedly when asking these questions for similar-but-distinct types.

I'll file individual issues for each analysis and offer to mentor them. They would be good bugs for people who have contributed before and might be interested in more.

File: src/ir/template.rs (hunk 7 above)
Comment (fitzgen) [inline src/ir/template.rs]: I kind of want to do a renaming pass "named" -> "type parameter" everywhere as a follow up. What do you think of that instead? (So this would become `AsTypeParameter` or `AsTypeParam`).

File: tests/expectations/lib.rs (hunk 8 above)
Comment (fitzgen) [inline tests/expectations/lib.rs]: Partially. I still think we want to insert this into all the bindings all the time.

Also, for some reason I still see the warnings *only some of the time* in the expectations tests, and I can't figure out why... This is definitely an improvement, however.

**Comment by fitzgen** [comment]
> We could add an annotation that is like "please consider all my template parameters used" and it should be pretty easy. Unless it is crucial for Stylo, I'd like to do such things as a follow up.

And we could deny this annotation/flag/option on type aliases, which is where the correctness issues come into play.

File: src/ir/context.rs (hunk 6 above)
Comment (fitzgen) [inline src/ir/context.rs]: Done

File: src/ir/context.rs [hunk 12]
```diff
@@ -306,6 +312,32 @@ impl<'ctx> BindgenContext<'ctx> {
...
     }
 
+    /// Add a new named template type parameter to this context's item set.
+    pub fn add_named_type(&mut self, item: Item, definition: clang::Cursor) {
+        debug!("BindgenContext::add_named_type: item = {:?}; definition = {:?}",
+               item,
+               definition);
+
+        assert!(item.is_named());
+        assert!(definition.kind() == clang_sys::CXCursor_TemplateTypeParameter);
```
Comment (fitzgen) [inline src/ir/context.rs]: Done

File: src/ir/context.rs [hunk 13]
```diff
@@ -306,6 +312,32 @@ impl<'ctx> BindgenContext<'ctx> {
...
+
+        let old_named_ty = self.named_types.insert(definition, id);
+        assert!(old_named_ty.is_none(),
+                "should not have already associated a named type with this id");
+    }
+
+    /// Get the named type defined at the given cursor location, if we've
+    /// already added one.
+    pub fn get_named_type(&self, definition: &clang::Cursor) -> Option<ItemId> {
+        assert!(definition.kind() == clang_sys::CXCursor_TemplateTypeParameter);
```
Comment (fitzgen) [inline src/ir/context.rs]: Done.

File: src/ir/named.rs (hunk 9 above)
Comment (fitzgen) [inline src/ir/named.rs]: Done.

File: src/ir/named.rs (hunk 10 above)
Comment (emilio) [inline src/ir/named.rs]: > It will only run in debug builds, which I don't think anyone is using outside hacking on bindgen itself, so I think we're fine.

You forget everyone doing a debug build with build-time bindgen, unfortunately :(

File: src/ir/named.rs [hunk 14]
```diff
@@ -327,10 +327,11 @@ impl<'ctx, 'gen> MonotoneFramework for UsedTemplateParameters<'ctx, 'gen>  {
                     &TypeKind::TemplateInstantiation(ref inst) => {
                         let decl = ctx.resolve_type(inst.template_definition());
                         let args = inst.template_arguments();
+                        // Although template definitions should always have
+                        // template parameters, there is a single exception:
+                        // opaque templates. Hence the unwrap_or.
```
Comment (fitzgen) [inline src/ir/named.rs]: This ends up being a bit gross to assert because of typedefs having different names than the underlying type, needing to see through resolved type references, and the fact that either of the template instantiation or the template definition can be opaque.

Unless you feel very strongly about it, I'm going to skip this assert for now.

File: src/ir/named.rs (hunk 10 above)
Comment (fitzgen) [inline src/ir/named.rs]: Ugh, you're right :(

I think we should probably define our own `testing_assert!(...)` macro that only runs under tests and replace **all** of our debug asserts with that.

Follow up?

File: src/ir/ty.rs (hunk 4 above)
Comment (fitzgen) [inline src/ir/ty.rs]: The last commit regarding partial template specializations obviates the need for this.

File: src/ir/named.rs (hunk 10 above)
Comment (emilio) [inline src/ir/named.rs]: That sounds good.

File: src/ir/item.rs (hunk 3 above)
Comment (emilio) [inline src/ir/item.rs]: Ok, I'm fine with that, thanks for doing it! :)

**Comment by fitzgen** [comment]
For posterity, here is the `creduce`d version of the failing stylo bindings:

```c++
template <typename> class a;
template <typename b, typename... c> class a<b(c...)> { a(const a &); };
template <typename b, typename... c> a<b(c...)>::a(const a &) {}
```

I really hope that this isn't `creduce` "kindly" figuring out that we don't support anonymous template parameters again, and skipping past the real issue I was trying to pin down... Although I do have a branch with a fix for the anonymous type parameters that I should be able to rebase pretty quickly.

**Comment by fitzgen** [comment]
Small update:

I fixed that test case, but am `creduce`ing another failure when generating Stylo bindings.

I think the fully pre-processed Stylo bindings are even bigger than SpiderMonkeys, and `creduce` is taking almost 24h to come up with reductions on Stylo bindings :-/

I guess the bright side is that it's good to get more edge cases into our test suite so we won't have to do this as often in the future!

**Comment by fitzgen** [comment]
Next stylo-derived creduced test case:

```c++
template <typename a> struct b {
  template <typename> using c = typename a::d;
  template <typename e> using f = b<c<e>>;
  f<int> g;
};
```

**Comment by fitzgen** [comment]
Grr, we're generating infinitely sized types for the most recent creduced test case.

**Comment by fitzgen** [comment]
@emilio OK, this should be ready for final review now.

Fixing that last test case was pretty nasty. We can't do anything sane for associated types other than recognize them and treat them as opaque, but of course clang doesn't expose the type, so we have to do some nasty regex matching on the type's spelling :(

**Comment by fitzgen** [comment]
CI is green :)

File: src/ir/template.rs [hunk 15]
```diff
@@ -91,12 +91,13 @@ impl TemplateInstantiation {
...
-            .filter(|t| t.kind() != CXType_Invalid)
-            .map(|t| Item::from_ty_or_ref(t, t.declaration(), None, ctx))
-            .collect();
+        let template_args = ty.template_args()
+            .map(|args| {
+                args.filter(|t| t.kind() != CXType_Invalid)
+                    .map(|t| Item::from_ty_or_ref(t, t.declaration(), None, ctx))
+                    .collect()
+            })
+            .unwrap_or(vec![]);
```
Comment (emilio) [inline src/ir/template.rs]: nit: `map_or`?

File: src/clang.rs [hunk 16]
```diff
@@ -913,6 +914,30 @@ impl Type {
...
+    /// this example:
+    ///
+    /// ```c++
+    /// template <typename T>
+    /// class Foo {
+    ///     typename T::Associated member;
+    /// };
+    /// ```
+    pub fn is_associated_type(&self) -> bool {
+        // This is terrible :(
```
Comment (emilio) [inline src/clang.rs]: ;_;

**Comment by emilio** [comment]
r=me. The regex thing makes me a bit uneasy, but not that I have a better solution, so... Please squash everything or remove the "Fix bad rebase" commit before landing :)

**Comment by fitzgen** [comment]
Fixed final nitpicks and squashed. Thanks for the review!

@bors-servo r=emilio

**Comment by emilio** [comment]
Wohoo!
